diff -urNp coreutils-6.11-orig/tests/cp/thru-dangling coreutils-6.11/tests/cp/thru-dangling
--- coreutils-6.11-orig/tests/cp/thru-dangling	2008-04-19 23:34:23.000000000 +0200
+++ coreutils-6.11/tests/cp/thru-dangling	2008-04-23 16:52:30.000000000 +0200
@@ -21,6 +21,7 @@ if test "$VERBOSE" = yes; then
   cp --version
 fi
 
+. $srcdir/../lang-default
 . $srcdir/../test-lib.sh
 
 ln -s no-such dangle || framework_failure
diff -urNp coreutils-6.11-orig/tests/chmod/thru-dangling coreutils-6.11/tests/chmod/thru-dangling
--- coreutils-6.11-orig/tests/chmod/thru-dangling	2008-04-19 23:34:23.000000000 +0200
+++ coreutils-6.11/tests/chmod/thru-dangling	2008-04-23 16:52:05.000000000 +0200
@@ -21,6 +21,7 @@ if test "$VERBOSE" = yes; then
   chmod --version
 fi
 
+. $srcdir/../lang-default
 . $srcdir/../test-lib.sh
 
 ln -s non-existent dangle || framework_failure
diff -urNp coreutils-6.11-orig/tests/misc/printf-surprise coreutils-6.11/tests/misc/printf-surprise
--- coreutils-6.11-orig/tests/misc/printf-surprise	2008-04-19 23:34:23.000000000 +0200
+++ coreutils-6.11/tests/misc/printf-surprise	2008-04-23 16:53:09.000000000 +0200
@@ -23,6 +23,8 @@ if test "$VERBOSE" = yes; then
   "$prog" --version
 fi
 
+
+. $srcdir/../lang-default
 . $srcdir/../test-lib.sh
 require_ulimit_
 
