diff -urNp coreutils-6.12-orig/src/seq.c coreutils-6.12/src/seq.c
--- coreutils-6.12-orig/src/seq.c	2008-05-26 08:40:32.000000000 +0200
+++ coreutils-6.12/src/seq.c	2008-09-29 22:09:21.000000000 +0200
@@ -304,7 +304,10 @@ print_numbers (char const *fmt, struct l
 	      bool print_extra_number = false;
 	      long double x_val;
 	      char *x_str;
-	      int x_strlen = asprintf (&x_str, fmt, x);
+	      int x_strlen;
+	      setlocale (LC_NUMERIC, "C");
+	      x_strlen = asprintf (&x_str, fmt, x);
+	      setlocale (LC_NUMERIC, ""); 
 	      if (x_strlen < 0)
 		xalloc_die ();
 	      x_str[x_strlen - layout.suffix_len] = '\0';
